### PR TITLE
Add copilot custom instructions

### PR DESCRIPTION
**Pull request by marcos-iov**
This pull request introduces comprehensive documentation to guide contributors and reviewers working on the powpeg-node project. It adds a Copilot custom instructions file tailored to the repository and establishes a set of Rootstock-specific coding principles. These documents emphasize security, reviewability, and maintainability, especially in critical code paths.

The most important changes are:

**Repository guidance and security practices:**
* Added `.github/copilot-instructions.md` to provide Copilot and human contributors with detailed repository context, security-critical areas, PR review priorities, build/test commands, and sensitive areas requiring extra scrutiny.
* Defined strict rules for handling secrets, configuration, and behavioral changes in security-sensitive code, with clear instructions for reviewers and contributors.

**Coding and style standards:**
* Added `docs/coding-principles.md`, documenting Rootstock-specific coding principles for naming, functions, classes, abstractions, error handling, comments, and testing. These guidelines prioritize clarity, maintainability, and correctness, especially for blockchain and money-handling code.
* Provided Java-specific style guidance, including naming conventions, use of enums, constructor injection, and handling of nullable values.

**Testing and PR process:**
* Established clear expectations for test coverage, test isolation, and CI workflows, including which files and behaviors require extra tests and scrutiny. [[1]](diffhunk://#diff-227c2c26cb2ee0ce0f46a320fc48fbcbdf21801a57f59161b1d0861e8aad55f5R1-R144) [[2]](diffhunk://#diff-f38733490899df0b900751348aa175dc4903ffff644afcd95c4b92ffa0a92571R1-R156)

### CODE REVIEW

File: .github/copilot-instructions.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+## Security-critical code
+
+powpeg-node participates in Rootstock's PowPeg infrastructure and handles Bitcoin transactions, signatures, and federation operations.
+
+When working in security-sensitive areas:
+
+* Security and correctness are more important than elegance.
+* Preserve existing behavior unless the task explicitly requires a behavioral change.
+* Do not refactor signing, peg-out, federation, key-management, transaction-construction, or consensus-related code unless required by the task.
+* Prefer proven, explicit code over clever abstractions in consensus- or money-related paths.
```
Comment (julia-zack) [inline .github/copilot-instructions.md]:
```suggestion
* Prefer proven, explicit code over clever abstractions in consensus or money-related paths.
```

File: .github/copilot-instructions.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+Reviewability is the top priority.
+
+* Keep changes scoped to the stated motivation. Flag broad, unrelated, or hard-to-review diffs.
+* Do not mix formatting-only changes with behavioral changes in the same commit.
+* Treat changes to signing, peg-out, Bitcoin connectivity, Bridge interaction, federation logic, key handling, and HSM integration as security-sensitive.
+* New behavior requires tests, including edge cases and error paths.
+* Avoid flaky tests: no Thread.sleep synchronization, real network calls without stubs, real-clock assumptions, fragile filesystem assumptions, or hidden dependency on test ordering.
+* Do not introduce secrets, private keys, real signatory keys, node keys, RPC credentials, or production config values into source, tests, docs, or logs.
+* Be conservative with logging around signatures, private/public key material, raw transactions, Bridge calls, and Bitcoin/Rootstock RPC data.
```
Comment (julia-zack) [inline .github/copilot-instructions.md:33]: maybe this should be in the Security-critical code section?

File: .github/copilot-instructions.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+* Code is read more often than written. Optimize for readability and maintainability.
+* Use intention-revealing names.
+* Include units in monetary and time-related names, such as amountInSatoshis, amountInWeis, amountInRBTC, and timeoutMillis.
+* Never rely on implicit monetary units.
+* Prefer focused functions and classes with clear responsibilities.
+* Eliminate duplication when practical.
+* Prefer self-explanatory code over explanatory comments.
+* Tests should be readable, independent, and cover boundary/error cases.
+* AI-generated refactorings must improve clarity, maintainability, or correctness. Do not introduce abstractions, indirection, or code movement without a clear benefit.
```
Comment (julia-zack) [inline .github/copilot-instructions.md]:
```suggestion
* Refactorings must improve clarity, maintainability, or correctness. Do not introduce abstractions, indirection, or code movement without a clear benefit.
```
prob all refactors should follow this :)

File: .github/copilot-instructions.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+## RSKj dependency workflow
+
+powpeg-node depends on co.rsk:rskj-core. For local development against a local RSKj checkout, use DONT-COMMIT-settings.gradle; never commit this file.
+
+The settings file may include a local composite build that substitutes co.rsk:rskj-core when the requested version ends in SNAPSHOT or RC.
+
+## CI gates a reviewer must predict
+
+Workflows live in .github/workflows/.
```
Comment (julia-zack) [inline .github/copilot-instructions.md:83]:
```suggestion
Workflows live in .github/workflows/
```
?

File: .github/copilot-instructions.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+Apply extra scrutiny to changes involving:
+
+* BTC, RSK, or MST signer configuration.
+* key-file handling, HSM handling, signing, signature serialization, and transaction signing.
+* peg-out transaction construction, validation, broadcasting, or retry behavior.
+* Bitcoin peer configuration and P2P/RPC assumptions.
+* Bridge contract interaction and RSKj integration points.
+* thread lifecycle, shutdown behavior, scheduling, retries, and network error handling.
+* dependency versions for rskj-core, bitcoinj, Jackson, logging, and cryptographic/signing libraries.
```
Comment (julia-zack) [inline .github/copilot-instructions.md:119]: 10000000% nit:

```suggestion
* dependency versions for rskj-core, bitcoinj, jackson, logging, and cryptographic/signing libraries.
```

File: .github/copilot-instructions.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+* packages: lowercase
+* classes: UpperCamelCase
+* methods and fields: lowerCamelCase
+* constants: CONSTANT_CASE
+
+Prefer constructor injection and immutable fields for new code when practical. Validate required constructor arguments. Avoid nullable returns where Optional<T> would make behavior clearer.
+
+## Testing guidance
+
+Unit-test new logic close to the changed class. Mock/stub Bitcoin, Rootstock, filesystem, clock, and network boundaries unless the test is explicitly an integration test.
```
Comment (julia-zack) [inline .github/copilot-instructions.md]: should we add that we should avoid mocks for other things as much as possible?

File: docs/coding-principles.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+## General principles
+
+* Code is read far more often than it is written. Optimize for readability and maintainability.
+* Leave touched code in a better state than you found it.
+* Small improvements are valuable: rename unclear variables, remove duplication, simplify conditions, and delete dead code when safe.
+* Clarity is more important than cleverness.
+* Avoid introducing unnecessary dependencies.
+* Prefer boring, obvious code over surprising code.
+* Code should look like it was written by someone who cared.
```
Comment (julia-zack) [inline docs/coding-principles.md:16]: ❤️

File: docs/coding-principles.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+Names are one of the most important parts of readability.
+
+* Use intention-revealing names.
+* Prefer names that reveal intent without requiring comments.
+* Names should describe domain concepts, not implementation details.
+* Include units in names when working with monetary values, time values, or measurements, such as amountInSatoshis, amountInWeis, amountInRBTC, timeoutMillis, and blockHeight.
+* Never rely on implicit monetary units.
+* Avoid vague suffixes such as Data, Info, Manager, Processor, or Helper unless they accurately describe the abstraction.
+* Avoid names that differ only in small or unclear ways.
+* Prefer searchable names over overly short names.
```
Comment (julia-zack) [inline docs/coding-principles.md:30]:
```suggestion
* Prefer searchable names over short names.
```
🤔

File: docs/coding-principles.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+* Class names with vague words such as Manager, Processor, Super, Data, or Info may indicate unclear responsibilities.
+* Prefer composition over inheritance.
+* Base classes should not know implementation details of derived classes.
+* Keep interfaces small and focused.
+* Keep coupling low by limiting what each class exposes.
+* Hide data, utility functions, constants, and temporary details when possible.
+* Avoid weakening encapsulation just to make testing easier unless there is no better option.
+* Public constants should come first, followed by private static fields, then private instance fields.
+* Public methods should generally appear before private helpers.
+* Private helpers may be placed near the public method that uses them to preserve readability.
```
Comment (julia-zack) [inline docs/coding-principles.md:74]:
```suggestion
* Public methods should generally appear before private helpers, but private helpers may be placed near the public method that uses them to preserve readability.
```
wdyt?

File: docs/coding-principles.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+* Comments should explain intent, rationale, protocol constraints, security considerations, or non-obvious decisions.
+* Comments should say what the code cannot say for itself.
+* Do not add comments that merely restate the code.
+* Do not use comments to compensate for unclear code when the code can be improved instead.
+* Remove obsolete comments and commented-out code.
+* Do not add author attribution comments; Git history already provides this information.
+* Do not add journal-style comments describing historical changes.
+* A useful comment should be accurate, clear, and maintained with the code.
+
+## Testing
```
Comment (julia-zack) [inline docs/coding-principles.md:123]: wdyt about adding a "prefer TDD approach when possible" point?

File: docs/coding-principles.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+* Avoid excessive test cleverness.
+
+## Pull requests and refactoring
+
+* Keep changes focused and reviewable.
+* Avoid large refactorings unrelated to the goal of the change.
+* Do not mix formatting-only changes with behavioral changes.
+* If a cleanup would significantly obscure the purpose of a PR, perform it separately.
+* AI-generated suggestions are not automatically improvements. Evaluate them critically.
+* AI-generated refactorings must improve clarity, maintainability, or correctness.
```
Comment (julia-zack) [inline docs/coding-principles.md]: same here
```suggestion
* Refactorings must improve clarity, maintainability, or correctness.
```

**Review by julia-zack — COMMENTED** [review]
looking very good, left some minor suggestions

File: .github/copilot-instructions.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+* Prefer proven, explicit code over clever abstractions in consensus- or money-related paths.
+* Prioritize behavior preservation over stylistic improvements.
+* Refactor security-sensitive code only when there is a clear benefit and adequate test coverage.
+* Be conservative when changing validation, serialization, cryptographic, or transaction-building logic.
+
+## PR review priorities
+
+Reviewability is the top priority.
+
+* Keep changes scoped to the stated motivation. Flag broad, unrelated, or hard-to-review diffs.
```
Comment (julianlen) [inline .github/copilot-instructions.md:27]: not sure what this means

File: docs/coding-principles.md (hunk 8 above)
Comment (julianlen) [reply docs/coding-principles.md:30]: I guess here it means "excessively"

File: docs/coding-principles.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+* Use intention-revealing names.
+* Prefer names that reveal intent without requiring comments.
+* Names should describe domain concepts, not implementation details.
+* Include units in names when working with monetary values, time values, or measurements, such as amountInSatoshis, amountInWeis, amountInRBTC, timeoutMillis, and blockHeight.
+* Never rely on implicit monetary units.
+* Avoid vague suffixes such as Data, Info, Manager, Processor, or Helper unless they accurately describe the abstraction.
+* Avoid names that differ only in small or unclear ways.
+* Prefer searchable names over overly short names.
+* Prefer pronounceable names.
+* Use one word for one concept and use it consistently.
```
Comment (julianlen) [inline docs/coding-principles.md:32]: also to keep consistency with what is already written

File: docs/coding-principles.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+* Keep variables close to where they are used.
+* Replace magic numbers with named constants.
+* Remove unused methods.
+
+## Classes
+
+* Classes should have a single responsibility and a single reason to change.
+* Prefer many focused classes over large multi-purpose classes.
+* Avoid "god classes".
+* Class names with vague words such as Manager, Processor, Super, Data, or Info may indicate unclear responsibilities.
```
Comment (julianlen) [inline docs/coding-principles.md:65]:
```suggestion
* Avoid Class names with vague words such as Manager, Processor, Super, Data, or Info, it may indicate unclear responsibilities.
```

more prompt-ish?

File: docs/coding-principles.md [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+* Robustness and readability are not conflicting goals.
+
+## Comments
+
+* Prefer self-explanatory code over explanatory comments.
+* Comments should explain intent, rationale, protocol constraints, security considerations, or non-obvious decisions.
+* Comments should say what the code cannot say for itself.
+* Do not add comments that merely restate the code.
+* Do not use comments to compensate for unclear code when the code can be improved instead.
+* Remove obsolete comments and commented-out code.
```
Comment (julianlen) [inline docs/coding-principles.md:118]: what about the TODOs dispersed around the codebase?

**Review by julianlen — COMMENTED** [review]
Great! Here are more instructions to consider
- Avoid having more than one unnecessary blank space.
- Never assign a constant to a variable
- Always leave one blank line at the end of every file.
- Remove anything that is not used: variables, functions, imports.
- Use functional style whenever possible, but avoid large-scale refactoring to functional code in a single PR.
- Keep consistency with the rest of the codebase whenever possible

Perhaps having a Codestyle section?

File: docs/coding-principles.md (hunk 8 above)
Comment (julia-zack) [reply docs/coding-principles.md:30]: good point

File: .github/copilot-instructions.md (hunk 2 above)
Comment (marcos-iov) [reply .github/copilot-instructions.md:33]: Could be, it is also a PR review priority. Let's just see how it goes

File: .github/copilot-instructions.md (hunk 4 above)
Comment (marcos-iov) [reply .github/copilot-instructions.md:83]: `.` means current dir, so it's basically the same thing

File: .github/copilot-instructions.md (hunk 5 above)
Comment (marcos-iov) [reply .github/copilot-instructions.md:119]: It's the library name, usually spelled with upper case

File: docs/coding-principles.md (hunk 9 above)
Comment (marcos-iov) [reply docs/coding-principles.md:74]: Not sure, but I think it works best if we put it as separate bullet points. At least that's what I read

File: docs/coding-principles.md (hunk 10 above)
Comment (marcos-iov) [reply docs/coding-principles.md:123]: Not sure it can be detected in a PR if tests were written first or afterwards

File: .github/copilot-instructions.md (hunk 12 above)
Comment (marcos-iov) [reply .github/copilot-instructions.md:27]: No big PRs changing many different unrelated things

File: docs/coding-principles.md (hunk 15 above)
Comment (marcos-iov) [reply docs/coding-principles.md:118]: Let's see how it behaves and then we can iterate

File: docs/coding-principles.md (hunk 14 above)
Comment (marcos-iov) [reply docs/coding-principles.md:65]: Not sure it makes a difference, let's see how it goes

File: docs/coding-principles.md (hunk 10 above)
Comment (julia-zack) [reply docs/coding-principles.md:123]: i guess only if tests are committed first. But nevermind, it'd make sense for a copilot coder eventually

**Review by julia-zack — APPROVED** [review]

**Review by julianlen — APPROVED** [review]
